test/Microsoft.DotNet.Tools.Tests.Utilities/Commands/DotnetCommand.cs

```diff
@@ -15,6 +17,41 @@ public DotnetCommand()
         public DotnetCommand(string dotnetUnderTest)
             : base(dotnetUnderTest)
         {
+            // Some scripts (e.g. RunCsc/RunVbc) depend on the correct `dotnet`
+            // being in the PATH. When tests use `DotnetUnderTest.WithBackwardsCompatibleRuntimes`,
+            // this means that the up-to-date dotnet command is resolved by the
+            // old-runtime-targeting RunCsc/Roslyn, causing compatability issues.
+            // So, detect if we're running a dotnet that's not just a plain "dotnet",
+            // and if so, make sure it's on the PATH.
+            var containingDirectory = Path.GetDirectoryName(dotnetUnderTest);
+            if (!string.IsNullOrEmpty(containingDirectory))
+            {
+                // if it's already set, use that
+                string path;
+                if (!Environment.TryGetValue("PATH", out path))
+                {
+                    // otherwise grab it from the environment
```

Comment:
There is too much comments. And people have to first understand the comments and do the code change. please build up abstraction by chain of responsibility pattern or at least extract several methods and put the comment as their name.